hbase-client/src/main/java/org/apache/hadoop/hbase/client/HMasterAsyncRegistry.java

```diff
@@ -0,0 +1,236 @@
+        TimeUnit.MILLISECONDS.toNanos(conf.getLong(HBASE_RPC_TIMEOUT_KEY,
+            DEFAULT_HBASE_RPC_TIMEOUT)));
+  }
+
+  private void parseHortPorts() {
+    String hostPorts = conf.get(CONF_KEY, DEFAULT_HOST_PORT);
+    for (String hostPort: hostPorts.split(",")) {
+      masterServers.add(ServerName.valueOf(hostPort, ServerName.NON_STARTCODE));
+    }
+    Preconditions.checkArgument(!masterServers.isEmpty(), String.format("%s is empty", CONF_KEY));
+    // Randomize so that not every client sends requests in the same order.
+    Collections.shuffle(masterServers);
+  }
+
+  /**
+   * Util that generates a master stub for a given ServerName.
+   */
+  private MasterService.BlockingInterface getMasterStub(ServerName server) throws IOException  {
+    return MasterService.newBlockingStub(
+        rpcClient.createBlockingRpcChannel(server, User.getCurrent(), rpcTimeout));
+  }
+
+  /**
+   * Blocking RPC to fetch the meta region locations using one of the masters from the parsed list.
+   */
+  private RegionLocations getMetaRegionLocationsHelper() throws MetaRegionsNotAvailableException {
+    List<HBaseProtos.RegionLocation> result = null;
+    for (ServerName sname: masterServers) {
+      try {
+        MasterService.BlockingInterface stub = getMasterStub(sname);
+        HBaseRpcController rpcController = RpcControllerFactory.instantiate(conf).newController();
+        GetMetaLocationsResponse resp = stub.getMetaLocations(rpcController,
+            GetMetaLocationsRequest.getDefaultInstance());
+        result = resp.getLocationsList();
+      } catch (Exception e) {
+        LOG.warn("Error fetch meta locations from master {}. Trying others.", sname, e);
+      }
+    }
+    if (result == null || result.isEmpty()) {
+      throw new MetaRegionsNotAvailableException(String.format(
+          "Meta locations not found. Probed masters: %s", conf.get(CONF_KEY, DEFAULT_HOST_PORT)));
+    }
+    List<HRegionLocation> deserializedResult = new ArrayList<>();
+    result.stream().forEach(
+        location -> deserializedResult.add(ProtobufUtil.toRegionLocation(location)));
+    return new RegionLocations(deserializedResult);
+  }
+
+  /**
+   * Picks the first master entry from 'masterHortPorts' to fetch the meta region locations.
+   */
+  @Override
+  public CompletableFuture<RegionLocations> getMetaRegionLocation() {
+    CompletableFuture<RegionLocations> result = new CompletableFuture<>();
+    CompletableFuture.runAsync(() -> {
+      try {
+        result.complete(this.getMetaRegionLocationsHelper());
+      } catch (Exception e) {
+        result.completeExceptionally(e);
+      }
+    });
+    return result;
+  }
+
+  /**
+   * Blocking RPC to get the cluster ID from the parsed master list. Returns null if no active
+   * master found.
+   */
+  private String getClusterIdHelper() throws  MasterNotRunningException {
+    // Loop through all the masters serially. We could be hitting some standby masters which cannot
```

Comment:
It's worth asking: Why do we need an active master to tell us what the cluster ID is? All the masters can look at the file in HDFS regardless of role. Standbys aren't special here, they can only serve one cluster, just like the active. If we fix this so any master can respond to this query then it's a small improvement in overall availability.

Reply:
Thats a good point. I'm looking into it.